src/sage/groups/finitely_presented.py

```diff
@@ -794,11 +772,45 @@ def __init__(self, free_group, relations, category=None):
         assert isinstance(relations, tuple)
         self._free_group = free_group
         self._relations = relations
-        self._assign_names(free_group.variable_names())
-        parent_gap = free_group.gap() / libgap([rel.gap() for rel in relations])
-        ParentLibGAP.__init__(self, parent_gap)
+        try:
+            self._assign_names(free_group.variable_names())
+        except ValueError:
+            pass
+        if libgap_fpgroup is None:
+            libgap_fpgroup = free_group.gap() / libgap([rel.gap() for rel in relations])
+        ParentLibGAP.__init__(self, libgap_fpgroup)
         Group.__init__(self, category=category)
 
+    def __hash__(self):
+        """
+        Make hashable.
+
+        EXAMPLES::
+
+            sage: G = FreeGroup(2) / [(1, 2, 2, 1)]
+            sage: G.__hash__() == hash((G.free_group(), G.relations()))
+            True
+        """
+        return hash((self._free_group, self._relations))
+
+    def __richcmp__(self, other, op):
+        """
+        Rich comparison of ``self`` and ``other``.
+
+        EXAMPLES::
+
+            sage: G1 = FreeGroup(2) / [(1, 2, 2, 1, 2, 1)]
+            sage: G2 = FreeGroup(2) / [(1, 2, 2, 1, 2, 1)]
+            sage: G1 == G2
+            True
+        """
```

Comment:
```suggestion
            True
            sage: G1 is G2
            False
        """
```

Reply:
Actually I got `True` for `G1 is G2`

Reply:
Then currently this test is not good. It should be comparing equal-but-not-identical groups. (Actually, we might want to have a shortcut for identical groups, but that might be a premature optimization.)

Reply:
You are right, the groups `G1` and `G2` are defined exactly in the same way. I tried several ways to produce `True` for `==` and `False` for `is` with other definitions but I failed.

Reply:
How about passing the gap group?

Reply:
It worked defining `G2 = libgap(G1).sage()`